onnx/defs/generator/defs.cc

```diff
@@ -523,4 +527,188 @@ ONNX_OPERATOR_SET_SCHEMA(
+      // we want max(0, ceil_cast_int) as negative values would evaluate to bool true in next step
+      {{"ceil_cast_int_relu"}, "Relu", {"ceil_cast_int"}},
+      {{"ceil_cast_bool"}, "Cast", {"ceil_cast_int_relu"}, {{"to", static_cast<int64_t>(9)}}},
+      {{"variadic_output", "output"}, "Loop", {"ceil_cast_int_relu", "ceil_cast_bool", "start"}, {MakeAttribute("body", loop_sub_graph)}}
+  });
+
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    Range,
+    11,
+    OpSchema()
+        .SetDoc(Range_ver11_doc)
+        .Input(
+            0,
+            "start",
+            "Scalar. First entry in the range.",
+            "T")
+        .Input(
+            1,
+            "limit",
+            "Scalar. Upper limit of sequence, exclusive.",
+            "T")
+        .Input(
+            2,
+            "delta",
+            "(Optional) Scalar. Number that increments start. Defaults to 1 if not provided.",
+            "T",
+            OpSchema::Optional)
+        .Output(
+            0,
+            "output",
+            "A 1-D tensor with same type as the inputs containing generated sequence.",
+            "T")
+        .TypeConstraint(
+            "T",
+            {"tensor(float)",
+             "tensor(double)",
+             "tensor(int16)",
+             "tensor(int32)",
+             "tensor(int64)"},
+            "Constrain input types to common numeric type tensors.")
+        .FunctionBody(build_nodes_range_op())
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          // Type inference
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+          // Shape inference
+          auto num_inputs = ctx.getNumInputs();
+
+          const auto* start_initializer = ctx.getInputData(0);
+          const auto* limit_initializer = ctx.getInputData(1);
+          const auto* delta_initializer =
+              num_inputs == 3 ? ctx.getInputData(2) : nullptr;
+
+          // Output is always 1-D
+          auto* output_dim = ctx.getOutputType(0)
+                                 ->mutable_tensor_type()
+                                 ->mutable_shape()
+                                 ->add_dim();
+
+          // all three inputs (if num_inputs == 3) (or)
+          // all two inputs (if num_inputs == 2)
+          // need to be available to infer output dim value
+          if (start_initializer != nullptr && limit_initializer != nullptr &&
+              (num_inputs == 2 || delta_initializer != nullptr)) {
+            if (start_initializer->data_type() !=
+                    limit_initializer->data_type() ||
+                (delta_initializer != nullptr &&
+                 start_initializer->data_type() !=
+                     delta_initializer->data_type())) {
+              fail_shape_inference(
+                  "All inputs to 'Range' op must be of the same type");
+            }
+
+            int output_dim_value = -1;
```

Comment:
```suggestion
            // If any of Range's inputs are not initializers, the output dimension would be unknown. 
            int output_dim_value = -1;
```

Reply:
Thanks for this comment. But I felt this comment is more suitable at the place we do the check in line 680 (the if... there does this check). The comment is added there. let me know what you think of this.

Reply:
Both work to me. If you move it to line 680, please do
```suggestion
            // If any of Range's inputs are not initializers, the output dimension would be unknown (denoted by -1 below). 
            int output_dim_value = -1;
```

Reply:
Moved already. Resolved.

Reply:
Just to make sure: why initialize this to -1 here? Please do not set_dim_value to -1: if it is unknown, please leave dim_value not set. It doesn't look like this -1 will reach the set_dim_value, but I just want to make sure. This is a slight problem because ONNX uses a convention slightly different from several frameworks where a single integer is used to represent both known and unknown dimensions, and this seems to confuse people.